eclipse/che
# Handle the mixed endpoints correctly with singlehost strategy

### What does this PR do?
This PR makes sure that both the workspace servers conforming to the server exposure strategy as well as the servers requiring subdomains work correctly under singlehost.

### Screenshot/screencast of this PR
N/A


### What issues does this PR fix or reference?
#18065 

### How to test this PR?
1. Create a workspace using the devfile below (notice that one of the endpoints is defined as `secure` and `cookiesAuthEnabled`)
1. Start the workpace
1. Start the server using the "Start Development mode" command
1. Click on the "hello-greeting-endpoint" and notice that the runtime token is loaded and the page is correctly displayed.
1. Repeat the above steps with che server running in `multi-host`, `single-host` and `single-host` with `singleHostExposureStrategy` set to `gateway` (the default under OpenShift)

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Additional Info

The devfile for testing:

```yaml
metadata:
  name: quarkus-cp088
projects:
  - name: quarkus-quickstarts
    source:
      location: 'https://github.com/quarkusio/quarkus-quickstarts.git'
      type: git
      sparseCheckoutDir: getting-started
components:
  - id: redhat/quarkus-java11/latest
    type: chePlugin
  - mountSources: true
    endpoints:
      - name: hello-greeting-endpoint
        port: 8080
        attributes:
          secure: 'true'
          cookiesAuthEnabled: 'true'
    memoryLimit: 4927M
    type: dockerimage
    volumes:
      - name: m2
        containerPath: /home/user/.m2
    image: 'quay.io/eclipse/che-quarkus:nightly'
    alias: centos-quarkus-maven
    env:
      - value: >-
          -XX:MaxRAMPercentage=50.0 -XX:+UseParallelGC -XX:MinHeapFreeRatio=10
          -XX:MaxHeapFreeRatio=20 -XX:GCTimeRatio=4
          -XX:AdaptiveSizePolicyWeight=90 -Dsun.zip.disableMemoryMapping=true
          -Xms20m -Djava.security.egd=file:/dev/./urandom -Duser.home=/home/user
        name: JAVA_OPTS
      - value: $(JAVA_OPTS)
        name: MAVEN_OPTS
  - mountSources: true
    endpoints:
      - name: hello-greeting-endpoint
        port: 8080
        attributes:
          path: /hello/greeting/che-user
    command:
      - tail
    args:
      - '-f'
      - /dev/null
    memoryLimit: 32M
    type: dockerimage
    image: registry.access.redhat.com/ubi8/ubi-minimal
    alias: ubi-minimal
apiVersion: 1.0.0
commands:
  - name: Package
    actions:
      - workdir: '${CHE_PROJECTS_ROOT}/quarkus-quickstarts/getting-started'
        type: exec
        command: ./mvnw package
        component: centos-quarkus-maven
  - name: Start Development mode (Hot reload + debug)
    actions:
      - workdir: '${CHE_PROJECTS_ROOT}/quarkus-quickstarts/getting-started'
        type: exec
        command: './mvnw compile quarkus:dev'
        component: centos-quarkus-maven
  - name: Package Native
    actions:
      - workdir: '${CHE_PROJECTS_ROOT}/quarkus-quickstarts/getting-started'
        type: exec
        command: ./mvnw package -Dnative -Dmaven.test.skip
        component: centos-quarkus-maven
  - name: Start Native
    actions:
      - workdir: '${CHE_PROJECTS_ROOT}/quarkus-quickstarts/getting-started/target'
        type: exec
        command: ./getting-started-1.0-SNAPSHOT-runner
        component: ubi-minimal
  - name: Attach remote debugger
    actions:
      - referenceContent: |
          {
            "version": "0.2.0",
            "configurations": [
              {
                "type": "java",
                "request": "attach",
                "name": "Attach to Remote Quarkus App",
                "hostName": "localhost",
                "port": 5005
              }
            ]
          }
        type: vscode-launch
```
